# What is Polymorphism?

Polymorphism means “many forms”. 

In Python (and OOP in general), it refers to the ability of different classes to respond to the same method call in different ways.

# Real-World Analogy:

When we call start():

1. A Car starts its engine

2. A Laptop powers on

3. A Microwave begins heating

All respond to start() — but behave differently.

### This is polymorphism.

# Types of Polymorphism in Python

| Type                         | Description                                                             |
| ---------------------------- | ----------------------------------------------------------------------- |
| **1. Duck Typing**           | Type checking is done at runtime based on behavior, not inheritance     |
| **2. Method Overriding**     | Same method in parent and child class, but child provides its own logic |
| **3. Function Polymorphism** | Built-in functions behave differently based on the input type           |
| **4. Operator Overloading**  | Same operator (`+`, `*`, etc.) used with different types                |


# 1. Duck Typing (Dynamic Polymorphism)

"If it walks like a duck and quacks like a duck, it's a duck."

Python is dynamically typed — it doesn't care about the object’s class, just whether it has the right methods.

In [1]:
class PyCharm:
    def execute(self):
        print("Compiling and running code in PyCharm.")

class VSCode:
    def execute(self):
        print("Running code with debugger in VS Code.")

# Polymorphic function
def run_code(ide):
    ide.execute()




In [2]:
# Object create
ide1 = PyCharm()
ide2 = VSCode()

run_code(ide1)  # PyCharm behavior
run_code(ide2)  # VS Code behavior

Compiling and running code in PyCharm.
Running code with debugger in VS Code.


### No inheritance is used, but run_code() works on both — this is duck typing.

### What's Happening Here?
# Step 1: Two Different Classes

We define two classes:

1. PyCharm

2. VSCode

### Each class has a method called execute(), but each method behaves differently.

# Step 2: A Polymorphic Function

### We create a function:

In [3]:
def run_code(ide):
    ide.execute()


### This function takes any object as ide and calls execute() on it.

# We can Notice:

1. It doesn’t care what class the object is from.

2. It only assumes the object has an execute() method.

# Step 3: Calling the Function with Different Objects

In [4]:
ide1 = PyCharm()
ide2 = VSCode()

run_code(ide1)
run_code(ide2)


Compiling and running code in PyCharm.
Running code with debugger in VS Code.


# Now, we're passing different objects (ide1, ide2) to the same function run_code().

### What happens?

1. **When run_code(ide1) runs** ide1.execute() calls the execute() in PyCharm

2. **When run_code(ide2) runs:** ide2.execute() calls the execute() in VSCode

3. Each object responds differently to the same method call — **that’s polymorphism.**

# Why is this called "Duck Typing"?

“If it looks like a duck, swims like a duck, and quacks like a duck — it’s a duck.”

### In Python:

1. If an object has an execute() method, we don’t care what class it is from.
( Python doesn’t require the object to be from a particular class or inherit from a specific base class.
We can pass any object to a function, as long as that object has the method being called (execute() in this case).)

2. We only care that it can respond to execute().

This is Duck Typing – Python doesn’t check the type, it checks the behavior at runtime.

# Why This Is Polymorphism

| Feature                   | Explanation                                                                |
| ------------------------- | -------------------------------------------------------------------------- |
| **Same interface**        | Both classes use the method name `execute()`                               |
| **Different behavior**    | `PyCharm.execute()` and `VSCode.execute()` behave differently              |
| **Unified function**      | `run_code()` calls `execute()` without caring about the actual object type |
| **No inheritance needed** | Python allows this through dynamic typing                                  |


# 2. Method Overriding (Runtime Polymorphism)

Subclass provides a specific implementation of a method already defined in its parent class.

In [5]:
class Employee:
    def role(self):
        print("General Employee")

class Programmer(Employee):
    def role(self):
        print("Writes code and builds software")

class Manager(Employee):
    def role(self):
        print("Manages project and team")




In [6]:
# Object creation
e1 = Employee()
e2 = Programmer()
e3 = Manager()



In [7]:
# Call
e1.role()
e2.role()
e3.role()

General Employee
Writes code and builds software
Manages project and team


# All objects respond to the same role() method, but differently based on their class.

 # Breakdown
 We have a Parent class: Employee
➤ It defines a method role()

We have two Child Class : Programmer and Manager
➤ Both inherit from Employee
➤ Both redefine the role() method

# Why is this Method Overriding?

Let’s focus on the method role():
| Class        | Method Name | Behavior                                    |
| ------------ | ----------- | ------------------------------------------- |
| `Employee`   | `role()`    | Prints: `"General Employee"`                |
| `Programmer` | `role()`    | Prints: `"Writes code and builds software"` |
| `Manager`    | `role()`    | Prints: `"Manages project and team"`        |


# All classes have a method with the same name: role()

But:

The base class Employee provides a generic definition.

The child classes Programmer and Manager provide specific versions of that method.

When we do:

In [8]:
e2 = Programmer()
e2.role()  # Calls the overridden method in Programmer


Writes code and builds software


### Python ignores the base class version and calls the child class’s version.

# Why it's important

This allows different subclasses to behave differently for the same interface (role()), which is a key feature of:

Polymorphism

Dynamic method dispatch

# Summary:
| Feature              | Your Code                                              |
| -------------------- | ------------------------------------------------------ |
| Parent method        | `Employee.role()`                                      |
| Overridden in child  | `Programmer.role()`, `Manager.role()`                  |
| Behavior change      | Yes — each prints a different message                  |
| Inheritance involved | Yes — both subclasses inherit from `Employee`          |
| Why it's overriding  | Same method name, different implementation in subclass |


# Another Example

In [9]:
## Parent Class
class Animal:
    def speak(self):
        return "Sound of the animal"
    
## Derived Class 1
class Dog(Animal):
    def speak(self):
        return "Woof!"
    
## Derived class
class Cat(Animal):
    def speak(self):
        return "Meow!"
    
## Function that demonstrates polymorphism
def animal_speak(animal):
    print(animal.speak())
    
dog=Dog()
cat=Cat()
print(dog.speak())
print(cat.speak())
animal_speak(dog)

Woof!
Meow!
Woof!


# Let's analyze each class:
✅ Parent class:

In [10]:
class Animal:
    def speak(self):
        return "Sound of the animal"


Method speak() returns a generic animal sound.

This is the base version of the method.

### ✅ Child class Dog overrides speak():

In [11]:
class Dog(Animal):
    def speak(self):
        return "Woof!"


### Dog inherits from Animal

It has its own speak() method — same name, but different behavior.

This is method overriding.

### ✅ Child class Cat overrides speak():

In [12]:
class Cat(Animal):
    def speak(self):
        return "Meow!"


### Same thing here — Cat overrides speak() with a different implementation.

### ✅ What Happens When we Call speak()?

In [13]:
print(dog.speak())  # Woof!
print(cat.speak())  # Meow!


Woof!
Meow!


# Python looks at the object’s actual class:

dog is a Dog, so it calls the Dog version of speak().

cat is a Cat, so it calls the Cat version of speak().

Even though both inherit from Animal, Python ignores the base class version if the method is overridden in the child class.

# What is happening here 

In [14]:
# Function that demonstrates polymorphism
def animal_speak(animal):
    print(animal.speak())

# Calling through polymorphic function
animal_speak(dog)


Woof!


# 1. animal_speak(animal) is a general-purpose function.

a. It accepts any object as the argument (parameter animal).

b. It then calls animal.speak() on that object.

c. It doesn’t care what class the object belongs to, as long as it has a method called speak()

# 2. dog is an instance of the Dog class.

Earlier in our code, we did:

In [15]:
dog = Dog()


### And in the Dog class, you overrode the speak() method:

In [16]:
class Dog(Animal):
    def speak(self):
        return "Woof!"


### So dog.speak() will return "Woof!".

# 3. When we call:

In [17]:
animal_speak(dog)


Woof!


### It executes:

In [18]:
print(dog.speak())  # Internally from the Dog class


Woof!


# Why This is Polymorphism
Polymorphism means:

The same interface (method name) behaves differently depending on the object it is acting upon.

In your function, you're always calling .speak()

But the actual method being run depends on whether you pass in a Dog, Cat, or any other subclass of Animal

✅ So:

| Object passed | Method executed  | Output                |
| ------------- | ---------------- | --------------------- |
| `Dog()`       | `Dog.speak()`    | `Woof!`               |
| `Cat()`       | `Cat.speak()`    | `Meow!`               |
| `Animal()`    | `Animal.speak()` | `Sound of the animal` |


# 3. Functions and Methods in Python
Polymorphism with functions and methods means:

Functions or methods can work with different types of data or objects, and respond differently depending on the input type or class — but using the same name.

### Two Key Forms:
| Type                      | Description                                                          |
| ------------------------- | -------------------------------------------------------------------- |
| **Function Polymorphism** | Same function name behaves differently depending on the input type   |
| **Method Polymorphism**   | Same method name behaves differently depending on the object’s class |


### 1️⃣ Function Polymorphism

### ➤ Definition:

A function behaves differently based on the type of input arguments, even though the function name is the same.

### ✅ Example 1: Built-in len() function

In [19]:
print(len("Hello"))       # Output: 5
print(len([1, 2, 3]))     # Output: 3
print(len({'a': 1, 'b': 2}))  # Output: 2


5
3
2


### Even though we're calling the same function len(), the behavior differs:
| Input Type | Behavior                     | Result |
| ---------- | ---------------------------- | ------ |
| `str`      | Returns number of characters | `5`    |
| `list`     | Returns number of elements   | `3`    |
| `dict`     | Returns number of keys       | `2`    |


### ✅ Same function name → different behavior depending on type.

### ✅ Example 2: User-defined function working on different types

In [20]:
def add(a, b):
    return a + b

print(add(5, 3))          # Integer addition → 8
print(add("Hi ", "there"))# String concatenation → "Hi there"
print(add([1], [2]))      # List merging → [1, 2]


8
Hi there
[1, 2]


| Input Type | Behavior             | Result       |
| ---------- | -------------------- | ------------ |
| Integers   | Adds numbers         | `8`          |
| Strings    | Concatenates strings | `"Hi there"` |
| Lists      | Merges lists         | `[1, 2]`     |


### ✅ Same function add() → polymorphic behavior.

# 2️⃣ Method Polymorphism
### ➤ Definition:

Different classes define the same method name, but with different behaviors depending on the object.

This is usually achieved by method overriding in subclasses.

In [21]:
class Animal:
    def speak(self):
        return "Some generic animal sound"

class Dog(Animal):
    def speak(self):
        return "Woof!"

class Cat(Animal):
    def speak(self):
        return "Meow!"

# Polymorphic function
def animal_sound(animal):
    print(animal.speak())

# Usage
dog = Dog()
cat = Cat()

animal_sound(dog)   # Woof!
animal_sound(cat)   # Meow!


Woof!
Meow!


# What's happening:

All classes have a method called speak()

Each class implements it differently

The function animal_sound() uses polymorphism to handle any animal

### Using Loops to Demonstrate Polymorphism with Methods

In [22]:
animals = [Dog(), Cat(), Animal()]

for animal in animals:
    print(animal.speak())


Woof!
Meow!
Some generic animal sound


In [32]:
### Polymorphissm with Functions and MEthods
## base class
class Shape:
    def area(self):
        return "The area of the figure"
    
## Derived class 1
class Rectangle(Shape):
    def __init__(self,width,height):
        self.width=width
        self.height=height

    def area(self):
        return self.width * self.height
    
##DErived class 2

class Circle(Shape):
    def __init__(self,radius):
        self.radius=radius

    def area(self):
        return 3.14*self.radius *self.radius
    
## Fucntion that demonstrates polymorphism

def print_area(shape):
    print(f"the area is {shape.area()}")


rectangle=Rectangle(4,5)
circle=Circle(3)

print_area(rectangle)
print_area(circle)

the area is 20
the area is 28.259999999999998


### Here we can see where we are overloading the function polimorphism

### Difference Between Function and Method Polymorphism
| Feature        | Function Polymorphism                    | Method Polymorphism                      |
| -------------- | ---------------------------------------- | ---------------------------------------- |
| Applies to     | Standalone functions                     | Class methods                            |
| Based on       | Type of arguments passed                 | Type of object on which method is called |
| Example        | `len()`, `add()`                         | `speak()` method in different classes    |
| Implemented by | Overloading built-in operators/functions | Overriding methods in subclasses         |


### Polymorphism using Abstract Base Classes (ABCs) in Python
This is a more formal and enforceable way to implement polymorphism, especially in large applications where code structure and consistency are important.

### One liner : Abstract class is an empty class and Abstract method is an empty method. We think that it can be used later , so whatever class is going to inherit this class will define this method of the class

### What is an Abstract Base Class (ABC)?

An abstract base class is a blueprint for other classes.

It cannot be instantiated directly.

It can define abstract methods that must be implemented by any subclass.

Enforces a consistent interface across all subclasses.

It's part of the abc module in Python.

### Why Use ABC for Polymorphism?

In regular polymorphism (like with Dog and Cat), we trust the child classes to override the method (speak()), but there’s no guarantee.

With an abstract base class, Python will raise an error if a child class doesn't implement the required methods.

🔐 ABC enforces polymorphism by design.

### How to Create an Abstract Base Class
Step-by-step:

Import ABC and abstractmethod from the abc module.

Create a class that inherits from ABC.

Use @abstractmethod to define abstract methods.

In [33]:
from abc import ABC, abstractmethod

# Abstract Base Class
class Shape(ABC):
    @abstractmethod #Abstract method
    def area(self):
        pass

# Subclass: Circle
class Circle(Shape):
    def __init__(self, radius):
        self.radius = radius

    def area(self):
        return 3.14 * self.radius * self.radius

# Subclass: Rectangle
class Rectangle(Shape):
    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height


In [24]:
shapes = [Circle(5), Rectangle(4, 6)]

for shape in shapes:
    print(f"Area: {shape.area()}")


Area: 78.5
Area: 24


### What If a Subclass Doesn’t Implement area()?

In [25]:
class Triangle(Shape):
    pass

t = Triangle()  # ❌ This will raise an error


TypeError: Can't instantiate abstract class Triangle without an implementation for abstract method 'area'

### That’s how ABCs force polymorphic behavior — if a subclass doesn’t implement the abstract method, Python won’t let you create an object from it.

### Polymorphism in Action

The key part is this loop:

In [26]:
for shape in shapes:
    print(shape.area())


78.5
24


In [34]:
from abc import ABC,abstractmethod

## Define an abstract class
class Vehicle(ABC):
    @abstractmethod
    def start_engine(self):
        pass

## Derived class 1
class Car(Vehicle):
    def start_engine(self):
        return "Car enginer started"
    
## Derived class 2
class Motorcycle(Vehicle):
    def start_engine(self):
        return "Motorcycle enginer started"
    
# Function that demonstrates polymorphism
def start_vehicle(vehicle):
    print(vehicle.start_engine())

## create objects of cAr and Motorcycle

car = Car()
motorcycle = Motorcycle()

start_vehicle(car)

Car enginer started


### shape could be any subclass of Shape

The method .area() will behave differently based on the actual object type

You’re using polymorphism safely and consistently — thanks to ABCs

# ✅ Real-World Analogy

Imagine we're building a drawing app:

We define a base Shape class (abstract).

We don’t care what kind of shape it is (circle, square, triangle).

We just call .area() or .draw() on every shape.

Using an ABC ensures that every shape object must implement those required behaviors.

| Feature                         | Explanation                                      |
| ------------------------------- | ------------------------------------------------ |
| Abstract Base Class (ABC)       | A blueprint for other classes                    |
| `@abstractmethod`               | Defines a method that must be overridden         |
| Cannot instantiate ABC directly | Enforces design rules                            |
| Used for                        | Structured, safe polymorphism                    |
| Benefits                        | Prevents runtime surprises, improves reliability |


## Bonus Example: Payment Gateway (Real-World Use Case)

In [27]:
from abc import ABC, abstractmethod

class PaymentMethod(ABC):
    @abstractmethod
    def pay(self, amount):
        pass

class CreditCard(PaymentMethod):
    def pay(self, amount):
        print(f"Paid ₹{amount} using Credit Card.")

class UPI(PaymentMethod):
    def pay(self, amount):
        print(f"Paid ₹{amount} via UPI.")

# Usage
def process_payment(payment: PaymentMethod, amount):
    payment.pay(amount)

payment1 = CreditCard()
payment2 = UPI()

process_payment(payment1, 1000)
process_payment(payment2, 500)


Paid ₹1000 using Credit Card.
Paid ₹500 via UPI.


### ✅ process_payment() works polymorphically with any valid PaymentMethod object.

# 4. Operator Overloading

### What is Operator Overloading?

Operator Overloading allows us to define custom behavior for built-in operators (+, -, *, etc.) when they are used with user-defined objects (instances of classes).

✅ It means: Same operator, different behavior depending on the operands’ types.

Just like functions can be polymorphic (work with different types), operators can be polymorphic too.

### Real-Life Analogy

+ for numbers: adds values → 5 + 3 = 8

+ for strings: concatenates → "hello" + "world" = "helloworld"

+ for lists: merges → [1,2] + [3,4] = [1,2,3,4]

 This is polymorphism of the + operator — and in custom classes, we can define this behavior ourselves using operator overloading.

### How to Overload Operators in Python

We overload an operator by defining special methods (dunder methods) in our class.

| Operator | Special Method             |
| -------- | -------------------------- |
| `+`      | `__add__(self, other)`     |
| `-`      | `__sub__(self, other)`     |
| `*`      | `__mul__(self, other)`     |
| `/`      | `__truediv__(self, other)` |
| `==`     | `__eq__(self, other)`      |
| `>`      | `__gt__(self, other)`      |
| `<`      | `__lt__(self, other)`      |
| `str()`  | `__str__(self)`            |


In [28]:
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __add__(self, other):
        return Point(self.x + other.x, self.y + other.y)

    def __str__(self):
        return f"({self.x}, {self.y})"


In [29]:
p1 = Point(1, 2)
p2 = Point(3, 4)

p3 = p1 + p2  # Internally calls p1.__add__(p2)

print(p3)     # Output: (4, 6)


(4, 6)


### Without __add__, Python wouldn’t know how to add two Point objects.

In [30]:
class Book:
    def __init__(self, title, pages):
        self.title = title
        self.pages = pages

    def __add__(self, other):
        return self.pages + other.pages

    def __eq__(self, other):
        return self.pages == other.pages

    def __gt__(self, other):
        return self.pages > other.pages

    def __str__(self):
        return f"{self.title} ({self.pages} pages)"


In [31]:
b1 = Book("Python Basics", 300)
b2 = Book("Advanced Python", 500)

print(b1 + b2)        # 800 → b1.__add__(b2)
print(b1 == b2)       # False → b1.__eq__(b2)
print(b2 > b1)        # True  → b2.__gt__(b1)


800
False
True


# Why Use Operator Overloading?
| Benefit                               | Explanation                                                                     |
| ------------------------------------- | ------------------------------------------------------------------------------- |
| Improves readability                  | Use natural syntax like `a + b` instead of `a.add(b)`                           |
| Makes classes act like built-in types | We can use arithmetic, comparison, and string functions on our custom objects |
| Enables polymorphism with operators   | Same operator can work across many types in different ways                      |


### Summary Table: Common Operator Methods
| Operator | Method Name   | Purpose               |
| -------- | ------------- | --------------------- |
| `+`      | `__add__`     | Addition              |
| `-`      | `__sub__`     | Subtraction           |
| `*`      | `__mul__`     | Multiplication        |
| `/`      | `__truediv__` | Division              |
| `==`     | `__eq__`      | Equal to              |
| `!=`     | `__ne__`      | Not equal to          |
| `>`      | `__gt__`      | Greater than          |
| `<`      | `__lt__`      | Less than             |
| `str()`  | `__str__`     | String representation |


In [37]:
### MAthematical operation for vectors
class Vector:
    def __init__(self,x,y):
        self.x=x
        self.y=y

    def __add__(self,other):
        return Vector(self.x+other.x,self.y+other.y)
    
    def __sub__(self, other):
        return Vector(self.x - other.x, self.y - other.y)

    def __mul__(self, other):
        return Vector(self.x * other, self.y * other)
    
    def __eq__(self, other):
        return self.x == other.x and self.y == other.y
    
    def __repr__(self):
        return f"Vector({self.x}, {self.y})"
    
## create objectss of the Vector Class
v1=Vector(2,3)
v2=Vector(4,5)


print(v1+v2)
print(v1-v2)
print(v1*3)

Vector(6, 8)
Vector(-2, -2)
Vector(6, 9)


# Let's Break It Down


In [39]:
class Vector:
    def __init__(self, x, y):
        self.x = x
        self.y = y


### This is the constructor, used to initialize a vector with 2 coordinates: x and y.


#### Overloading the + operator with __add__

In [40]:
def __add__(self, other):
    return Vector(self.x + other.x, self.y + other.y)


#### This allows us to do:

In [41]:
v1 + v2  # same as v1.__add__(v2)


✅ It adds two vectors component-wise:

x: 2 + 4 = 6

y: 3 + 5 = 8

Returns: Vector(6, 8)

### 2️⃣ Overloading the - operator with __sub__

In [42]:
def __sub__(self, other):
    return Vector(self.x - other.x, self.y - other.y)


### Now we can subtract two vectors:

In [43]:
v1 - v2  # same as v1.__sub__(v2)


# ✅ Subtracts each component:

x: 2 - 4 = -2

y: 3 - 5 = -2

Returns: Vector(-2, -2)

In [38]:
### Overloading Operators for Complex Numbers

class ComplexNumber:
    def __init__(self, real, imag):
        self.real = real
        self.imag = imag

    def __add__(self, other):
        return ComplexNumber(self.real + other.real, self.imag + other.imag)

    def __sub__(self, other):
        return ComplexNumber(self.real - other.real, self.imag - other.imag)

    def __mul__(self, other):
        real_part = self.real * other.real - self.imag * other.imag
        imag_part = self.real * other.imag + self.imag * other.real
        return ComplexNumber(real_part, imag_part)

    def __truediv__(self, other):
        denominator = other.real**2 + other.imag**2
        real_part = (self.real * other.real + self.imag * other.imag) / denominator
        imag_part = (self.imag * other.real - self.real * other.imag) / denominator
        return ComplexNumber(real_part, imag_part)

    def __eq__(self, other):
        return self.real == other.real and self.imag == other.imag

    def __repr__(self):
        return f"{self.real} + {self.imag}i"

# Create objects of the ComplexNumber class
c1 = ComplexNumber(2, 3)
c2 = ComplexNumber(1, 4)

# Use overloaded operators
print(c1 + c2)  # Output: 3 + 7i
print(c1 - c2)  # Output: 1 - 1i
print(c1 * c2)  # Output: -10 + 11i
print(c1 / c2)  # Output: 0.8235294117647058 - 0.29411764705882354i
print(c1 == c2) # Output: False

3 + 7i
1 + -1i
-10 + 11i
0.8235294117647058 + -0.29411764705882354i
False



### ✅ **Class Definition: `ComplexNumber`**

```python
class ComplexNumber:
```

We're defining a class to represent complex numbers.

---

### ✅ **Constructor: `__init__`**

```python
def __init__(self, real, imag):
    self.real = real
    self.imag = imag
```

* Initializes the real and imaginary parts of a complex number.
* No operator overloading here yet — just setting up the object’s internal state.

---

## 🎯 Now the operator overloading begins:

### ✅ **Overloading `+` Operator: `__add__`**

```python
def __add__(self, other):
    return ComplexNumber(self.real + other.real, self.imag + other.imag)
```

* Enables `c1 + c2`.
* Internally calls `c1.__add__(c2)`.
* Returns a new `ComplexNumber` with summed real and imaginary parts.

---

### ✅ **Overloading `-` Operator: `__sub__`**

```python
def __sub__(self, other):
    return ComplexNumber(self.real - other.real, self.imag - other.imag)
```

* Enables `c1 - c2` → calls `c1.__sub__(c2)`.

---

### ✅ **Overloading `*` Operator: `__mul__`**

```python
def __mul__(self, other):
    real_part = self.real * other.real - self.imag * other.imag
    imag_part = self.real * other.imag + self.imag * other.real
    return ComplexNumber(real_part, imag_part)
```

* Enables `c1 * c2` → calls `c1.__mul__(c2)`.
* Uses complex multiplication formula:

  $$
  (a + bi)(c + di) = (ac - bd) + (ad + bc)i
  $$

---

### ✅ **Overloading `/` Operator: `__truediv__`**

```python
def __truediv__(self, other):
    denominator = other.real**2 + other.imag**2
    real_part = (self.real * other.real + self.imag * other.imag) / denominator
    imag_part = (self.imag * other.real - self.real * other.imag) / denominator
    return ComplexNumber(real_part, imag_part)
```

* Enables `c1 / c2` → calls `c1.__truediv__(c2)`.
* Uses formula for complex division:

  $$
  \frac{a + bi}{c + di} = \frac{(ac + bd) + (bc - ad)i}{c^2 + d^2}
  $$

---

### ✅ **Overloading `==` Operator: `__eq__`**

```python
def __eq__(self, other):
    return self.real == other.real and self.imag == other.imag
```

* Enables `c1 == c2` → calls `c1.__eq__(c2)`.
* Compares if both real and imaginary parts are equal.

---

### ✅ **String Representation: `__repr__`**

```python
def __repr__(self):
    return f"{self.real} + {self.imag}i"
```

* Overloads how the object is **represented as a string**, e.g., in `print(c1 + c2)`.

---

## 🧪 **Now comes the usage part — where operator overloading is applied**

```python
c1 = ComplexNumber(2, 3)
c2 = ComplexNumber(1, 4)
```

* Two `ComplexNumber` objects are created.

---

### 🟢 `print(c1 + c2)`

* Internally calls `c1.__add__(c2)` due to operator overloading.
* Returns a new `ComplexNumber(3, 7)` → `3 + 7i`.

---

### 🟢 `print(c1 - c2)`

* Calls `c1.__sub__(c2)` → `1 - 1i`.

---

### 🟢 `print(c1 * c2)`

* Calls `c1.__mul__(c2)` using complex multiplication rules → `-10 + 11i`.

---

### 🟢 `print(c1 / c2)`

* Calls `c1.__truediv__(c2)` using complex division → `0.8235 - 0.2941i`.

---

### 🟢 `print(c1 == c2)`

* Calls `c1.__eq__(c2)` → compares the two → returns `False`.

---

### 🔚 Summary

Every operator (`+`, `-`, `*`, `/`, `==`) in the usage section triggers a corresponding method (`__add__`, `__sub__`, etc.) that you defined in the class. This is **operator overloading** — giving new meaning to standard Python operators when used with user-defined objects.

